drivers/can/can_xmc4xxx.c

```diff
@@ -0,0 +1,998 @@
+	reg |= FIELD_PREP(CAN_NODE_NBTR_SJW_Msk, timing->sjw - 1);
+
+	dev_cfg->can->NBTR = reg;
+
+	k_mutex_unlock(&dev_data->mutex);
+
+	return 0;
+}
+
+static int can_xmc4xxx_send(const struct device *dev, const struct can_frame *msg,
+			    k_timeout_t timeout, can_tx_callback_t callback, void *callback_arg)
+{
+	struct can_xmc4xxx_data *dev_data = dev->data;
+	uint8_t mailbox_idx;
+	struct can_xmc4xxx_tx_callback *callbacks = &dev_data->tx_callbacks[0];
+	CAN_MO_TypeDef *mo;
+	unsigned int key;
+
+	LOG_DBG("Sending %d bytes. Id: 0x%x, ID type: %s %s %s %s", can_dlc_to_bytes(msg->dlc),
+		msg->id, msg->flags & CAN_FRAME_IDE ? "extended" : "standard",
+		msg->flags & CAN_FRAME_RTR ? "RTR" : "",
+		msg->flags & CAN_FRAME_FDF ? "FD frame" : "",
+		msg->flags & CAN_FRAME_BRS ? "BRS" : "");
+
+	__ASSERT_NO_MSG(callback != NULL);
+
+	if (msg->dlc > CAN_XMC4XXX_MAX_DLC) {
+		return -EINVAL;
+	}
+
+	if (!dev_data->common.started) {
+		return -ENETDOWN;
+	}
+
+	if (dev_data->state == CAN_STATE_BUS_OFF) {
+		return -ENETUNREACH;
+	}
+
+	if ((msg->flags & (CAN_FRAME_FDF | CAN_FRAME_BRS)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if (k_sem_take(&dev_data->tx_sem, timeout) != 0) {
+		return -EAGAIN;
+	}
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	for (mailbox_idx = 0; mailbox_idx < CONFIG_CAN_XMC4XXX_MAX_TX_QUEUE; mailbox_idx++) {
+		if (callbacks[mailbox_idx].function == NULL) {
+			break;
+		}
+	}
+
+	__ASSERT_NO_MSG(mailbox_idx < CONFIG_CAN_XMC4XXX_MAX_TX_QUEUE);
+
+	key = irq_lock();
+	/* critical section in case can_xmc4xxx_reset_tx_fifos() called in isr  */
+	/* so that callback function and callback_arg are consistent */
+	callbacks[mailbox_idx].function = callback;
+	callbacks[mailbox_idx].user_data = callback_arg;
+	irq_unlock(key);
+
+	mo = dev_data->tx_mo[mailbox_idx];
+	mo->MOCTR = CAN_MO_MOCTR_RESMSGVAL_Msk;
+
+	if ((msg->flags & CAN_FRAME_IDE) != 0) {
+		/* MOAR - message object arbitration register */
+		mo->MOAR = FIELD_PREP(CAN_MO_MOAR_PRI_Msk, 1) |
+			   FIELD_PREP(CAN_MO_MOAR_ID_Msk, msg->id) | CAN_MO_MOAR_IDE_Msk;
+	} else {
+		mo->MOAR = FIELD_PREP(CAN_MO_MOAR_PRI_Msk, 1) |
+			   FIELD_PREP(XMC_CAN_MO_MOAR_STDID_Msk, msg->id);
+	}
+
+	mo->MOFCR &= ~CAN_MO_MOFCR_DLC_Msk;
+	mo->MOFCR |= FIELD_PREP(CAN_MO_MOFCR_DLC_Msk, msg->dlc);
+
+	if ((msg->flags & CAN_FRAME_RTR) != 0) {
+		mo->MOCTR = CAN_MO_MOCTR_RESDIR_Msk;
+	} else {
+		mo->MOCTR = CAN_MO_MOCTR_SETDIR_Msk;
+		memcpy((void *)&mo->MODATAL, &msg->data[0], sizeof(uint32_t));
+		memcpy((void *)&mo->MODATAH, &msg->data[4], sizeof(uint32_t));
+	}
+
+	mo->MOCTR = CAN_MO_MOCTR_SETTXEN0_Msk | CAN_MO_MOCTR_SETTXEN1_Msk |
+		    CAN_MO_MOCTR_SETMSGVAL_Msk | CAN_MO_MOCTR_RESRXEN_Msk |
+		    CAN_MO_MOCTR_RESRTSEL_Msk;
+	mo->MOCTR = CAN_MO_MOCTR_SETTXRQ_Msk;
+
+	k_mutex_unlock(&dev_data->mutex);
+	return 0;
+}
+
+static CAN_MO_TypeDef *can_xmc4xxx_get_mo(uint8_t *mo_index)
+{
+	int i;
+
+	for (i = 0; i < DT_PROP(MULTICAN_NODE, message_objects); i++) {
```

Comment:
Perhaps add a define for `DT_PROP(MULTICAN_NODE, message_objects)`? It's used in a couple of places.